libraries/GCS_MAVLink/GCS_Common.cpp

```diff
@@ -1743,6 +1743,22 @@ void GCS_MAVLINK::handle_common_gps_message(mavlink_message_t *msg)
     AP::gps().handle_msg(msg);
 }
 
+void GCS_MAVLINK::handle_system_time_message(mavlink_message_t *msg)
+{
+    // exit immediately if system time already set
+    if (hal.util->system_time_set()) {
+        return;
+    }
+
+    mavlink_system_time_t packet;
+    mavlink_msg_system_time_decode(msg, &packet);
+
+    // set system clock for log timestamps
+    hal.util->set_system_clock(packet.time_unix_usec);
+
+    // update signing timestamp
+    GCS_MAVLINK::update_signing_timestamp(packet.time_unix_usec);
```

Comment:
Pretty sure stuff will go to hell here if time goes backwards.

Reply:
If a board have a RTC, or anything that gives the correct time, `system_time_set` need to be true, like Linux boards.

Reply:
But the only other place that sets the system time in ardupilot is when the gps is updated, it has no relation to RTC at all.

And the fact that during initialization mavlink packets are being handled, chances are that if the GCS sends this packet it will set the time before the GPS set it.

What if the GCS wants to set the time again, shouldn't it be allowed?

This particular line here looks very wrong btw. If GCS tries to update the time multiple times, we would fail `hal.util->set_system_clock()` but change the signing timestamp.

Reply:
The GCS can look at sensor status flags and send **SYSTEM_TIME** if gps flags indicate that gps isn't connected.
Other alternative is to look at **SYSTEM_TIME** epoch time coming from autopilot
and set time only once if epoch time is zero _(happens when gps isn't connected)_.

This patch only allows to set time only once, the variable `system_time_was_set` will lock `set_system_clock` functionality after the first call. Besides that, it appears that this function only needs to be performed once, microsecond accuracy is not critical for `system_time`, right ?

In the ROV scenario, there are nearing thousands of sub users whose all logs are in 1970 because they cannot access 10k+ gps systems, this patch allows they to have a correct log time :bowtie:

Reply:
> What if the GCS wants to set the time again, shouldn't it be allowed?

No, it shouldn't be allowed. @patrickelectric is correct that the GCS will only be allowed to set the time once with this patch, and the signing timestamp will only be set once as a result.

Reply:
It looks like plane repeatedly sets the system time and updates the signing timestamp, unlike the other vehicles which currently do a check to make sure it's only done once (like this pr). Is this correct? https://github.com/ArduPilot/ardupilot/blob/master/ArduPlane/ArduPlane.cpp#L450

Reply:
this is still wrong. If GCS sends multiple times you `hal.util->set_system_clock(packet.time_unix_usec);` will fail but a) you will return all went well and b) you will call `GCS_MAVLINK::update_signing_timestamp(packet.time_unix_usec);`  that you shouldn't.

Reply:
@lucasdemarchi can you say where in the code this will happen ? The logic behind this is to set time once. 

> If GCS sends multiple times you hal.util->set_system_clock(packet.time_unix_usec); will fail but a) you will return all went well and b) you will call GCS_MAVLINK::update_signing_timestamp(packet.time_unix_usec); that you shouldn't.

I'll rewrite this.

Reply:
It's void, there is no return or ACK. The gcs should not care of the result.

> you will call GCS_MAVLINK::update_signing_timestamp(packet.time_unix_usec); that you shouldn't.

No, we are not. It happens one time only. And even if we did do it many times, _why_ shouldn't we? Plane calls this method repeatedly, so obviously it is ok.

Reply:
Yes, that's true. But GPS can do the same thing as the GCS, moving the time backwards. With the pixhawk usually it starts with 1970 or 2000, don't remember.